### ISSUE TITLE
"A decorator for 'I' was not provided with an instance parameter." when registering a Func that resolves from the ComponentContext

### ISSUE THREAD

**Issue by Pvlerick** [autofac/Autofac#1007]
```csharp
[Fact]
public void Test()
{
    var builder = new ContainerBuilder();

    builder.RegisterType<C>()
        .As<I>();

    builder.RegisterDecorator<I>(
        (_, __, inner) => new D(inner, "D"));

    //This registration will work
    //builder.RegisterType<C>()
    //    .Named<I>("C");

    //builder.RegisterDecorator<I>(
    //    (_, inner) => new D(inner, "D"),
    //    fromKey: "C");

    builder.Register<Func<I>>(ctx =>
    {
        var context = ctx.Resolve<IComponentContext>();

        return () =>
        {
            return context.Resolve<I>();
        };
    });

    var container = builder.Build();

    container.Resolve<Func<I>>()().M();
}

public interface I
{
    void M();
}

public class C : I
{
    public void M() => Console.WriteLine("C.M");
}

public class D : I
{
    private readonly I inner;
    private readonly string name;

    public D(I inner, string name)
    {
        this.inner = inner ?? throw new ArgumentNullException(nameof(inner));
        this.name = name ?? throw new ArgumentNullException(nameof(name));
    }

    public void M()
    {
        Console.WriteLine("D.M: {0}", this.name);
        this.inner.M();
    }
}
```
Autofact 4.9+

Full stacktrace:
```
Autofac.Core.DependencyResolutionException: An exception was thrown while activating λ:AutoFacSandbox.NewDecoratorSyntax+I.
   at Autofac.Core.Resolving.InstanceLookup.Activate(IEnumerable`1 parameters, Object& decoratorTarget) in C:\projects\autofac\src\Autofac\Core\Resolving\InstanceLookup.cs:line 135
   at Autofac.Core.Resolving.InstanceLookup.Execute() in C:\projects\autofac\src\Autofac\Core\Resolving\InstanceLookup.cs:line 83
   at Autofac.Core.Resolving.ResolveOperation.GetOrCreateInstance(ISharingLifetimeScope currentOperationScope, IComponentRegistration registration, IEnumerable`1 parameters) in C:\projects\autofac\src\Autofac\Core\Resolving\ResolveOperation.cs:line 131
   at Autofac.Core.Resolving.ResolveOperation.Execute(IComponentRegistration registration, IEnumerable`1 parameters) in C:\projects\autofac\src\Autofac\Core\Resolving\ResolveOperation.cs:line 84
   at lambda_method(Closure )
   at AutoFacSandbox.NewDecoratorSyntax.Test() in C:\Users\pvlerick\Source\Repos\junkcode\AutoFacSandbox\NewDecoratorSyntax.cs:line 75
Autofac.Core.DependencyResolutionException: A decorator for 'I' was not provided with an instance parameter.
   at Autofac.RegistrationExtensions.<>c__DisplayClass66_0`1.<RegisterDecorator>b__0(IComponentContext c, IEnumerable`1 p) in C:\projects\autofac\src\Autofac\RegistrationExtensions.cs:line 1388
   at Autofac.Builder.RegistrationBuilder.<>c__DisplayClass0_0`1.<ForDelegate>b__0(IComponentContext c, IEnumerable`1 p) in C:\projects\autofac\src\Autofac\Builder\RegistrationBuilder.cs:line 62
   at Autofac.Core.Activators.Delegate.DelegateActivator.ActivateInstance(IComponentContext context, IEnumerable`1 parameters) in C:\projects\autofac\src\Autofac\Core\Activators\Delegate\DelegateActivator.cs:line 71
   at Autofac.Core.Resolving.InstanceLookup.Activate(IEnumerable`1 parameters, Object& decoratorTarget) in C:\projects\autofac\src\Autofac\Core\Resolving\InstanceLookup.cs:line 117
```

**Comment by nblumhardt** [comment]
At first glance, does look like a bug 👍

Might be a good first issue for someone who's just joined the project?

**Comment by RaymondHuy** [comment]
Hi @nblumhardt  When I investigate on this issue, I also find this case:
```csharp
[Fact]
public void Test()
{
    var builder = new ContainerBuilder();

    builder.RegisterType<C>()
        .As<I>();
    builder.RegisterDecorator(typeof(D), typeof(I));

    builder.Register<Func<I>>(ctx =>
    {
        return () =>
        {
            return new B();
        };
    });

    var container = builder.Build();
    var instance = container.Resolve<Func<I>>()();
    Assert.IsType<B>(instance);  //Failed  Actually is D type
}
public class C : I
{
       public void M() => Console.WriteLine("C.M");
}
public class B: I
{
       public void M() => Console.WriteLine("B.M");
}
public class D : I
{
    private readonly I inner;
    public D(I inner)
    {
    }

    public void M()
    {
        this.inner.M();
    }
}
```
Is it the feature ? Because I think it makes sense to have B type instead of being decorated.

**Comment by nblumhardt** [comment]
Normally last-in-wins would apply here, and the `Func` registration would take precendence 👍  - I'm not sure whether that's the currently-expected behavior though, I'm basing that on old versions that I'm more familiar with... :-)

**Comment by nblumhardt** [comment]
Ah; because the decorator will be registered dynamically, it will in fact be "last", even though in syntax-order its first.. 🤔

**Comment by RaymondHuy** [comment]
So is it a bug ? If it is, I can try to fix it :)

**Comment by RaymondHuy** [comment]
Hi @Pvlerick  your expected result when you run 
```csharp
container.Resolve<Func<I>>()();
```
is that it will call the delegate
```csharp
var context = ctx.Resolve<IComponentContext>();
return () =>
{
 return context.Resolve<I>();
};
```
and return an instance with D type and it contains C type as inner. Is it right ? or you expect another output ?

**Comment by Pvlerick** [comment]
@RaymondHuy yes, that's what I expect. I basically expect the same behaviour as the one I'd get if I resolved I on the container.

**Comment by RaymondHuy** [comment]
Hi @nblumhardt I have made PR (https://github.com/autofac/Autofac/pull/1011) for this issue. Can you review it ? :)

**Comment by worming004** [comment]
Hi,
by removing the Register of Func<I>, there is no exception, and the resolution provides the simplest constructor 'as expected'.
`
[Fact]
        public void OriginalTest()
        {
            var builder = new ContainerBuilder();

            builder.RegisterType<C>()
                .As<I>();

            builder.RegisterDecorator<I>(
                (_, __, inner) => new D(inner, "D"));


            var container = builder.Build();

            container.Resolve<Func<I>>()().M(); // No exception, output prove that decorator and concrete method is called

        }
`

By debugging the @Pvlerick test, his own container.Resolve<Func<I>>() is not returning the register some lines above. It's still trying to resolve the decorator. See above 
`
[... same code as @Pvlerick  ...]
// We put again the register
 builder.Register<Func<I>>(ctx =>
    {
        var context = ctx.Resolve<IComponentContext>();

        return () =>
        {
            return context.Resolve<I>();
        };
    });
[...]
     var fu = container.Resolve<Func<I>>(); // resolve works, but is wrapped by Autofac's stuffes
     var inter = fu(); // exception is thrown here
     inter.M();
` 

In short, autofac's resolution is confused by  a register of a Func<interface> mixed with decorator.

**Comment by alexmg** [comment]
This should be resolved with #1023.
